pennylane/beta/plugins/default_qubit_tf.py

```diff
@@ -0,0 +1,442 @@
+        "passthru_interface": "tf",
+    }
+
+    operations = {
+        "BasisState",
+        "QubitStateVector",
+        "QubitUnitary",
+        "DiagonalQubitUnitary",
+        "PauliX",
+        "PauliY",
+        "PauliZ",
+        "MultiRZ",
+        "Hadamard",
+        "S",
+        "T",
+        "CNOT",
+        "SWAP",
+        "CSWAP",
+        "Toffoli",
+        "CZ",
+        "PhaseShift",
+        "RX",
+        "RY",
+        "RZ",
+        "Rot",
+        "CRX",
+        "CRY",
+        "CRZ",
+        "CRot",
+    }
+
+    observables = {"PauliX", "PauliY", "PauliZ", "Hadamard", "Hermitian", "Identity"}
+
+    parametric_ops = {
+        "PhaseShift": tf_ops.PhaseShift,
+        "RX": tf_ops.RX,
+        "RY": tf_ops.RY,
+        "RZ": tf_ops.RZ,
+        "Rot": tf_ops.Rot,
+        "CRX": tf_ops.CRX,
+        "CRY": tf_ops.CRY,
+        "CRZ": tf_ops.CRZ,
+    }
+
+    _asarray = staticmethod(tf.convert_to_tensor)
+
+    def __init__(self, wires, *, shots=1000, analytic=True):
+        # create the initial state
+        state = np.zeros(2 ** wires, dtype=np.complex128)
+        state[0] = 1
+        state = tf.convert_to_tensor(state, dtype=C_DTYPE)
+
+        # Internally, we store the state as a tensor of dimension [2]*wires
+        self._state = tf.reshape(state, [2] * wires)
+        self._pre_rotated_state = self._state
+
+        # call QubitDevice init
+        super().__init__(wires, shots, analytic)
+
+    def apply(self, operations, rotations=None, **kwargs):
+        rotations = rotations or []
+
+        # apply the circuit operations
+        for i, operation in enumerate(operations):
+
+            if i > 0 and isinstance(operation, (QubitStateVector, BasisState)):
+                raise DeviceError(
+                    "Operation {} cannot be used after other Operations have already been applied "
+                    "on a {} device.".format(operation.name, self.short_name)
+                )
+
+            self._apply_operation(operation)
+
+        # store the pre-rotated state
+        self._pre_rotated_state = self._state
+
+        # apply the circuit rotations
+        for operation in rotations:
+            self._apply_operation(operation)
+
+    def _apply_operation(self, operation):
+        """Applies operations to the internal device state.
+
+        Args:
+            operation (~.Operation): operation to apply to the device
+        """
+        if isinstance(operation, QubitStateVector):
+            self._apply_state_vector(operation.parameters[0], operation.wires)
+            return
+
+        if isinstance(operation, BasisState):
+            self._apply_basis_state(operation.parameters[0], operation.wires)
+            return
+
+        if isinstance(operation, DiagonalOperation):
+            if operation.name in self.parametric_ops:
+                matrix = self.parametric_ops[operation.name](*operation.parameters)
+                eigvals = tf.linalg.diag_part(matrix)

```

Comment:
I thought the same 🤔 Do you think it's worth it?

Reply:
Done! It was actually pretty simple; the relevant functions in `tf_ops` simply need to return a vector representing the diagonal, rather than the full matrix.